fdbserver/RestoreApplier.actor.cpp

```diff
@@ -152,313 +150,280 @@ ACTOR static Future<Void> handleSendMutationVectorRequest(RestoreSendVersionedMu
-	Version curTxnId;
-	Version uncommittedTxnId;
-	bool lastTxnHasError;
-
-	// Decide when to commit a transaction. We buffer enough mutations in a txn before commit the txn
-	bool startNextVersion; // The next txn will include mutations in next version
-	int numAtomicOps; // Status counter
-	double txnBytes; // Decide when to commit a txn
-	double txnMutations; // Status counter
-
-	Reference<ApplierBatchData> batchData;
-	UID applierId;
-
-	DBApplyProgress() = default;
-	explicit DBApplyProgress(UID applierId, Reference<ApplierBatchData> batchData)
-	  : applierId(applierId), batchData(batchData), curIndexInCurTxn(0), startIndexInUncommittedTxn(0), curTxnId(0),
-	    uncommittedTxnId(0), lastTxnHasError(false), startNextVersion(false), numAtomicOps(0), txnBytes(0),
-	    txnMutations(0) {
-		curItInCurTxn = batchData->kvOps.begin();
-		while (curItInCurTxn != batchData->kvOps.end() && curItInCurTxn->second.empty()) {
-			curItInCurTxn++;
+// Clear all ranges in input ranges
+ACTOR static Future<Void> applyClearRangeMutations(Standalone<VectorRef<KeyRangeRef>> ranges, Database cx) {
+	state Reference<ReadYourWritesTransaction> tr(new ReadYourWritesTransaction(cx));
+	loop {
+		try {
+			tr->reset();
+			tr->setOption(FDBTransactionOptions::ACCESS_SYSTEM_KEYS);
+			tr->setOption(FDBTransactionOptions::LOCK_AWARE);
+			for (auto& range : ranges) {
+				tr->clear(range);
+			}
+			wait(tr->commit());
+			break;
+		} catch (Error& e) {
+			wait(tr->onError(e));
 		}
-		startItInUncommittedTxn = curItInCurTxn;
 	}
+	return Void();
+}
 
-	// Has all mutations been committed?
-	bool isDone() { return curItInCurTxn == batchData->kvOps.end(); }
-
-	// Set cursor for next mutation
-	void nextMutation() {
-		curIndexInCurTxn++;
-		while (curItInCurTxn != batchData->kvOps.end() && curIndexInCurTxn >= curItInCurTxn->second.size()) {
-			curIndexInCurTxn = 0;
-			curItInCurTxn++;
-			startNextVersion = true;
+// Get keys in imcompleteStagingKeys and precompute the stagingKey which is stored in batchData->stagingKeys
+ACTOR static Future<Void> getAndComputeStagingKeys(
+    std::map<Key, std::map<Key, StagingKey>::iterator> imcompleteStagingKeys, Database cx, UID applierID) {
+	state Reference<ReadYourWritesTransaction> tr(new ReadYourWritesTransaction(cx));
+	state std::vector<Future<Optional<Value>>> fValues;
+	state std::vector<Optional<Value>> values;
+	state int i = 0;
+	TraceEvent("FastRestoreApplierGetAndComputeStagingKeysStart", applierID)
+	    .detail("GetKeys", imcompleteStagingKeys.size());
+	loop {
+		try {
+			tr->reset();
+			tr->setOption(FDBTransactionOptions::ACCESS_SYSTEM_KEYS);
+			tr->setOption(FDBTransactionOptions::LOCK_AWARE);
+			for (auto& key : imcompleteStagingKeys) {
+				fValues.push_back(tr->get(key.first));
+			}
+			for (i = 0; i < fValues.size(); i++) {
+				Optional<Value> val = wait(fValues[i]);
+				values.push_back(val);
+			}
```

Comment:
You can wait for all values are ready and then inserted them into values. This can avoid waiting sequentially.
```
wait(waitForAll(fValues));
std::transform(fValues.begin(), fValues.end(), std::back_inserter(values), [](const Future<Optional<Value>>& f) { return f.get(); });
```